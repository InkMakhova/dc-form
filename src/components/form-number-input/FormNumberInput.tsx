// import * as React from 'react'
import { JSX } from 'react'
import { Group, Input, Label, NumberField } from 'react-aria-components'
import PlusIcon from '../../assets/plus.svg?react'
import MinusIcon from '../../assets/minus.svg?react'
import inputStyles from '../form-input/FormInput.module.css'
import styles from './FormNumberInput.module.css'
import { ButtonSize, ButtonVariant } from '../form-button/FormButton.types.ts'
import { FormNumberInputProps } from './FormNumberInput.types.ts'
import { InputSize } from '../form-input/FormInput.types.ts'
import FormButton from '../form-button/FormButton.tsx'

function FormNumberInput(props: FormNumberInputProps): JSX.Element {
  const {
    name,
    label,
    placeholder,
    min,
    max,
    step,
    value,
    defaultValue,
    isDisabled,
    onChange
  } = props;

  return (
    <NumberField
      minValue={min}
      maxValue={max}
      step={step}
      value={value}
      defaultValue={defaultValue}
      className={inputStyles['container']}
      onChange={onChange}
    >
      <Label>{label}</Label>
      <Group className={styles['container']}>
        {/* Decrement Button */}
        <FormButton
          type="button"
          variant={ButtonVariant.SecondaryContained}
          size={ButtonSize.Small}
          icon={<MinusIcon />}
          ariaLabel="Decrease size"
          slot="decrement"
          isDisabled={isDisabled || Number(value) <= min}
        />
        {/* Number Input */}
        <Input
          name={name}
          placeholder={placeholder}
          className={`${inputStyles['input']} ${inputStyles[`input--${InputSize.Small}`]}`}
          disabled={isDisabled}
        />
        {/* Increment Button */}
        <FormButton
          type="button"
          variant={ButtonVariant.SecondaryContained}
          size={ButtonSize.Small}
          icon={<PlusIcon />}
          ariaLabel="Increase size"
          slot="increment"
          isDisabled={isDisabled || Number(value) >= max}
        />
      </Group>
    </NumberField>
  )
}

export default FormNumberInput;
