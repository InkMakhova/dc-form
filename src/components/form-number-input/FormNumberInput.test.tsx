import { fireEvent, render, screen } from '@testing-library/react';
import { expect, vitest} from 'vitest'
import FormNumberInput from './FormNumberInput.tsx';

test('renders input with label', () => {
  render(
    <FormNumberInput
      name="size"
      label="Size"
      value={5}
      min={1}
      max={10}
      step={1}
      onDecrement={() => {}}
      onIncrement={() => {}}
    />
  )
  expect(screen.getByText('Size')).toBeInTheDocument()
})

test('calls onIncrement when increment button is clicked', () => {
  const handleIncrement = vitest.fn()
  const handleDecrement = vitest.fn()
  render(
    <FormNumberInput
      name="size"
      label="Size"
      value={5}
      min={1}
      max={10}
      step={1}
      onIncrement={handleIncrement}
      onDecrement={handleDecrement}
    />
  )
  const incrementButton = screen.getByLabelText('Increase size')
  fireEvent.click(incrementButton)
  expect(handleIncrement).toHaveBeenCalled()
})

test('calls onDecrement when decrement button is clicked', () => {
  const handleDecrement = vitest.fn()
  const handleIncrement = vitest.fn()
  render(
    <FormNumberInput
      name="size"
      label="Size"
      value={5}
      min={1}
      max={10}
      step={1}
      onIncrement={handleIncrement}
      onDecrement={handleDecrement}
    />
  )
  const decrementButton = screen.getByLabelText('Decrease size')
  fireEvent.click(decrementButton)
  expect(handleDecrement).toHaveBeenCalled()
})

test('decrement button is disabled at min value', () => {
  const handleDecrement = vitest.fn()
  const handleIncrement = vitest.fn()

  render(
    <FormNumberInput
      name="size"
      label="Size"
      value={1}
      min={1}
      max={10}
      step={1}
      onIncrement={handleIncrement}
      onDecrement={handleDecrement}
    />
  )
  const decrementButton = screen.getByLabelText('Decrease size')
  expect(decrementButton).toBeDisabled()
})

test('increment button is disabled at max value', () => {
  const handleDecrement = vitest.fn()
  const handleIncrement = vitest.fn()
  render(
    <FormNumberInput
      name="size"
      label="Size"
      value={10}
      min={1}
      max={10}
      step={1}
      onIncrement={handleIncrement}
      onDecrement={handleDecrement}
    />
  )
  const incrementButton = screen.getByLabelText('Increase size')
  expect(incrementButton).toBeDisabled()
})

test('matches snapshot', () => {
  const handleDecrement = vitest.fn()
  const handleIncrement = vitest.fn()
  const { container } = render(
    <FormNumberInput
      name="size"
      label="Size"
      placeholder="Size"
      value={10}
      min={1}
      max={10}
      step={1}
      onIncrement={handleIncrement}
      onDecrement={handleDecrement}
    />
  )
  expect(container).toMatchSnapshot();
});
