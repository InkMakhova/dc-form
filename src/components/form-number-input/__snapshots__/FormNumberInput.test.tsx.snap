// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <div
    class="_input__container_7e3432"
    data-rac=""
  >
    <label
      class="react-aria-Label"
      for="react-aria-«r1s»"
      id="react-aria-«r1u»"
    >
      Size
    </label>
    <div
      class="_input-number__group-container_9885f5"
      data-rac=""
      role="group"
    >
      <button
        aria-controls="react-aria-«r1s»"
        aria-label="Decrease size"
        aria-labelledby="react-aria-«r23» react-aria-«r1u»"
        class="
        _button_4e4ded
        _button--small_4e4ded
        _button--secondary-contained_4e4ded
        _button--icon-only_4e4ded
        _button--decrement_4e4ded
      "
        data-rac=""
        data-testid="form-button"
        id="react-aria-«r23»"
        slot="decrement"
        tabindex="-1"
        type="button"
      >
        <span
          class="
            _button__icon_4e4ded 
            
          "
        >
          <svg
            fill="none"
            height="2"
            viewBox="0 0 14 2"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 1.99921V-0.000671387H13.9992V1.99921H0Z"
              fill="#1A1A1A"
            />
          </svg>
        </span>
      </button>
      <input
        aria-labelledby="react-aria-«r1u»"
        aria-readonly="true"
        aria-roledescription="Number field"
        autocomplete="off"
        autocorrect="off"
        class="
            _input__field_7e3432 
            _input__field--small_7e3432
            _input-number_9885f5
          "
        data-rac=""
        id="react-aria-«r1s»"
        inputmode="numeric"
        name="size"
        placeholder="Size"
        readonly=""
        spellcheck="false"
        tabindex="0"
        title=""
        type="text"
        value="10"
      />
      <button
        aria-controls="react-aria-«r1s»"
        aria-label="Increase size"
        aria-labelledby="react-aria-«r22» react-aria-«r1u»"
        class="
        _button_4e4ded
        _button--small_4e4ded
        _button--secondary-contained_4e4ded
        _button--icon-only_4e4ded
        _button--increment_4e4ded
      "
        data-disabled="true"
        data-rac=""
        data-testid="form-button"
        disabled=""
        id="react-aria-«r22»"
        slot="increment"
        tabindex="-1"
        type="button"
      >
        <span
          class="
            _button__icon_4e4ded 
            
          "
        >
          <svg
            fill="none"
            height="14"
            viewBox="0 0 14 14"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.49988 0.499695V6.49969H13.4989V7.49969H7.49988V13.4987H6.49988V7.49969H0.499878V6.49969H6.49988V0.499695H7.49988Z"
              fill="#1A1A1A"
              stroke="black"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
